### ISSUE TITLE
[YBASE] High latency for YSQL follower reads

### ISSUE THREAD

**Issue by nmantrav2** [yugabyte/yugabyte-db#6226]
Multi-region configuration. We are doing a read from a follower region (US west coast)... so read from leader (US east coast)  is expected to take 70ms... and read from follower (local region) should take ~1ms or less for this workload.

However, read-from-follower is taking 200-250ms... whereas read from leader is taking 70ms (as expected). 

Build # 2.3.3.0 # 106. 

```
postgres=# SET yb_read_from_followers = true;
SET
Time: 0.177 ms
postgres=# select * from postgresqlkeyvalue where k='key:1';
   k   |   v
-------+-------
 key:1 | val:1
(1 row)
Time: 185.782 ms
postgres=# select * from postgresqlkeyvalue where k='key:1';
   k   |   v
-------+-------
 key:1 | val:1
(1 row)
Time: 276.489 ms

VS

postgres=# SET yb_read_from_followers = false;
SET
Time: 0.220 ms
postgres=# select * from postgresqlkeyvalue where k='key:1';
   k   |   v
-------+-------
 key:1 | val:1
(1 row)
Time: 65.085 ms
postgres=# select * from postgresqlkeyvalue where k='key:1';
   k   |   v
-------+-------
 key:1 | val:1
(1 row)
Time: 65.228 ms
```

**Comment by kmuthukk** [comment]
Looked at this with @mbautin , and turning vlog level 3 on a follower yb-tserver indicated the following difference in the TServer level request for YSQL follower read vs. YCQL follower read

YSQL:
```
I1030 22:14:17.057010 27596 tablet_service.cc:1707] Received Read RPC: tablet_id: "7c29b1589ebe4f068d28c15f2fd47e16"
include_trace: false
consistency_level: CONSISTENT_PREFIX
propagated_hybrid_time: 6570377449704951808
read_time {
  read_ht: 6570377449704558592
  local_limit_ht: 6570377449909358592
  global_limit_ht: 6570377449909358592
  in_txn_limit_ht: 6570377449704849408
}
pgsql_batch {
  client: YQL_CLIENT_PGSQL
  stmt_id: 30529792
  schema_version: 0
  hash_code: 15703
  partition_column_values {
    value {
      string_value: "key:930342"
    }
  }
  targets {
    column_id: 0
  }
  targets {
    column_id: 1
  }
  targets {
    column_id: -8
  }
  column_refs {
    ids: 0
    ids: 1
  }
  is_forward_scan: true
  is_aggregate: false
  limit: 1024
  return_paging_state: true
  ysql_catalog_version: 6
  max_hash_code: 15703
  table_id: "000030ad000030008000000000004006"
}
proxy_uuid: ""
rejection_score: 0
batch_idx: 18446744073709551615
```

The YSQL request has a read time set (which should not be set for "read from follower" mode. And this is cause the follower to unnecessary wait for "safe time" to reach a certain value before it serves the read.

In contrast the YCQL requests (for read from follower mode) does not have the `read_time` field set.

```
I1030 22:14:17.211642 28366 tablet_service.cc:1707] Received Read RPC: tablet_id: "08eb59dfe3f64a0a8d60bf9f53581b25"
ql_batch {
  client: YQL_CLIENT_CQL
  request_id: 7988554785015037420
  schema_version: 0
  hash_code: 35137
  hashed_column_values {
    value {
      string_value: "1e6b739d-f9bf-4f58-b1ea-5f285fb931b9:509"
    }
  }
  selected_exprs {
    column_id: 0
  }
  selected_exprs {
    column_id: 1
  }
  limit: 5000
  return_paging_state: true
  distinct: false
  max_hash_code: 35137
  column_refs {
    ids: 0
    ids: 1
  }
  query_id: 57746688
  rsrow_desc {
    rscol_descs {
      name: "k"
      ql_type {
        main: STRING
      }
    }
    rscol_descs {
      name: "v"
      ql_type {
        main: BINARY
      }
    }
  }
  is_forward_scan: true
  is_aggregate: false
}
include_trace: false
consistency_level: CONSISTENT_PREFIX
propagated_hybrid_time: 6570377450338463744
proxy_uuid: "163bc67325374d698eccb918489690c9"
rejection_score: 0.49096547480257036
batch_idx: 18446744073709551615
```

**Comment by kmuthukk** [comment]
Additional note:

Explicitly setting the isolation level to READ COMMITTED (or READ COMMITTED, DEFERRABLE) brings down the latency, =

```
postgres=# SET yb_read_from_followers = true;
SET
Time: 0.180 ms
postgres=# select * from postgresqlkeyvalue where k = 'key:1';
   k   |   v
-------+-------
 key:1 | val:1
(1 row)
Time: 176.524 ms  <---- high (not expected)

postgres=# BEGIN ISOLATION LEVEL READ COMMITTED;
BEGIN
Time: 0.140 ms
postgres=# select * from postgresqlkeyvalue where k = 'key:1';
   k   |   v
-------+-------
 key:1 | val:1
(1 row)
Time: 0.788 ms   <---- seeing the expected low latency!
```

**Comment by kmuthukk** [comment]
Interestingly, when inside an active transaction, it seems to have the desired effect as opposed to a top-level SELECT. See below:

```
postgres=# SET yb_read_from_followers = false;
SET
Time: 0.163 ms

postgres=# BEGIN;
BEGIN
Time: 0.151 ms
postgres=# select * from postgresqlkeyvalue where k = 'key:1';
   k   |   v
-------+-------
 key:1 | val:1
(1 row)
Time: 0.893 ms  <--- fast
postgres=# commit;
COMMIT
Time: 0.137 ms
postgres=# select * from postgresqlkeyvalue where k = 'key:1';
   k   |   v
-------+-------
 key:1 | val:1
(1 row)
Time: 153.956 ms <--- slow
```

**Comment by kmuthukk** [comment]
@tedyu was looking at the above and commented that maybe this is the needed change:

```
diff --git a/src/postgres/src/backend/executor/nodeIndexscan.c b/src/postgres/src/backend/executor/nodeIndexscan.c
index 714c56fee..dd2255036 100644
--- a/src/postgres/src/backend/executor/nodeIndexscan.c
+++ b/src/postgres/src/backend/executor/nodeIndexscan.c
@@ -135,6 +135,7 @@ IndexNext(IndexScanState *node)
         scandesc->yb_exec_params = &estate->yb_exec_params;
         // Add row marks.
         scandesc->yb_exec_params->rowmark = -1;
+        scandesc->yb_exec_params->read_from_followers = YBReadFromFollowersEnabled();
         ListCell   *l;
         foreach(l, estate->es_rowMarks) {
             ExecRowMark *erm = (ExecRowMark *) lfirst(l);

```

**Comment by tedyu** [comment]
For what Kannan discovered, maybe this change is more relevant:
```
diff --git a/src/postgres/src/backend/executor/execUtils.c b/src/postgres/src/backend/executor/execUtils.c
index 7fcb8e20c..6704c409d 100644
--- a/src/postgres/src/backend/executor/execUtils.c
+++ b/src/postgres/src/backend/executor/execUtils.c
@@ -178,7 +178,7 @@ CreateExecutorState(void)
    estate->yb_exec_params.limit_use_default = true;
    estate->yb_exec_params.rowmark = -1;
    estate->yb_can_batch_updates = false;
-   estate->yb_exec_params.read_from_followers = false;
+   estate->yb_exec_params.read_from_followers = YBReadFromFollowersEnabled();

    return estate;
 }
```
This is for executor state creation.

**Comment by tedyu** [comment]
Performed the following operations using ysqlsh :
```
yugabyte=# SET yb_read_from_followers = true;
SET
yugabyte=# select count(*) from sales_p1;
 count
-------
     0
(1 row)

yugabyte=# select count(*) from t;
 count
-------
     2
(1 row)
```
Additional logging in CreateExecutorState showed the following:
```
2020-10-31 13:31:06.992 UTC [28737] LOG:  using read_from_followers: 0
2020-10-31 13:31:06.992 UTC [28737] STATEMENT:  SELECT n.nspname as "Schema",
...
2020-10-31 13:31:32.840 UTC [28737] LOG:  using read_from_followers: 1
2020-10-31 13:31:32.840 UTC [28737] STATEMENT:  select count(*) from sales_p1;
...
2020-10-31 13:31:46.792 UTC [28737] LOG:  using read_from_followers: 1
2020-10-31 13:31:46.792 UTC [28737] STATEMENT:  select count(*) from t;
```
We can see that read_from_followers was enabled in CreateExecutorState. Without patch, it was false.

**Comment by amitanandaiyer** [comment]
Will be fixed with #10473
